p2p: always provide CNodeStateStats and getpeerinfo/netinfo/gui updates

jonatack opened the pull request:
- update CNodeStateStats members to be std::optional or member-initialized with default values

- return CNodeStateStats and NodeStats directly rather than a bool

- make some getpeerinfo and gui peers fields always provided rather than optional: addr_processed, addr_rate_limited, addr_relay_enabled, inflight

- remove NodesStats type which causes more confusion than benefit

jonatack committed:
p2p, rpc, qt: always provide CNodeStateStats, update getpeerinfo and gui

- update CNodeStateStats members to be either std::optional, or member-initialized
  with default values where it makes sense and simplifies the code

- return CNodeStateStats rather than a bool

- in RPC getpeerinfo, continue to always provide the relaytxes and minfeefilter
  fields with default values of false and 0, respectively

- make getpeerinfo and gui peers fields always provided rather than optional where
  it makes sense: addr_processed, addr_rate_limited, addr_relay_enabled, inflight

jonatack committed:
scripted-diff: rename CNodeStateStats members

-BEGIN VERIFY SCRIPT-
rename() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

rename nSyncHeight      m_sync_height
rename nCommonHeight    m_common_height
rename vHeightInFlight  m_height_in_flight
rename presync_height   m_presync_height
rename their_services   m_services
-END VERIFY SCRIPT-

MarcoFalke commented:
Can you explain why this is a "bugfix", considering that the API break was done on purpose at that time?

vasild commented:
Concept ACK

@MarcoFalke, a bug is a bug, even if it was done on purpose. @jnewbery can clarify here, but I do not think this was even purposeful. My guess is that this breakage was unwanted, but acceptable in order to get the refactor from #21160 done. It later required fixup attempts like https://github.com/bitcoin/bitcoin/pull/24691 and https://github.com/bitcoin/bitcoin/pull/25176.

So, what happened? The `relaytxes` and `minfeefilter` fields were always present, and documented as such, in the output of the `getpeerinfo` RPC. #21160 changed it so that they could sometimes be missing. I guess the issues (both for GUI and RPC) were neglected during the review with a comment like [will almost always be true, so this isn't a big issue](https://github.com/bitcoin/bitcoin/pull/21160#discussion_r598567395). Well, it turned out our own tool `bitcoin-cli` broke due to the missing but expected fields. Some user apps will break too (if we don't fix it in 24.0).

https://github.com/bitcoin/bitcoin/pull/24691 documented the fields as optional, but that is not appropriate because in the previous version they were not-optional. When users wrote their apps they assumed the fields will always be present. I think the "optional" flag should only be added when a field is initially introduced. Adding it later is the same as dropping the field - it breaks the "contract" with the user that the field will be present.

On this PR: I think it is ok to display the values `relaytxes=false` and `minfeefilter=0` in those corner cases. This is what the code did before #21160. If this is stripped to the bare minimum needed for the fix, I think it will have higher chance of making it to 24.0

jnewbery commented:
> So, what happened? The relaytxes and minfeefilter fields were always present, and documented as such, in the output of the getpeerinfo RPC.

This isn't true. **None** of the fields in `getpeerinfo` were documented as always present until the 'optional' fields were explicitly documented in #22798. Prior to that PR, there was no explicit guarantee that any of the fields would be present.

PR 22798 was opened _after_ #21160, then merged, and resulted in the silent conflict where the fields were not marked as optional. That could be caught by better testing of optional fields.

> with a comment like "will almost always be true, so this isn't a big issue"

Stop misrepresenting other people's words. This was in response to a review comment about the GUI, not the RPC. You've also clipped the quote to misrepresent it further. Here's the full sentence: "will almost always be true, so this isn't a big issue (if it was, then the other stats in nodeStateStats not being available would also be an issue)." I stand by that - if it's a problem that one of the fields is not visible in the GUI for a very short window during peer connection, then it'd be a problem that the other fields like "startingheight" or "minping" are not present. It's clearly not a problem, since that's always been the case.

At the time of the comment, it was also totally consistent with the other fields that were optional but not marked as such, since #22798 had not yet been opened.

> in the previous version they were not-optional. When users wrote their apps they assumed the fields will always be present. I think the "optional" flag should only be added when a field is initially introduced. Adding it later is the same as dropping the field - it breaks the "contract" with the user that the field will be present.

Again, this is false. The 'optional' documentation was added in #22798, which was only included in v0.23. Until then, there were many fields could be considered optional but not documented as such. Users writing apps that interacted with the field (including the bitcoin-cli netinfo app) should have made those apps tolerant to the fields not being present.

As can be seen in #22798, there were 10s of fields across many different RPCs that were 'optional' before being explicitly documented as such. Both the PR description and @vasild's comment make it sound like this was some nefarious intentional API break, or some egregious bug. It's neither.

Frankly, it's utterly tedious to have to spend time rebutting false statements and misrepresentations made about PRs that were merged months ago. Unfortunately, this 'gotcha' attitude of trying to shame code authors seems to be becoming more common in this project. It's extremely harmful for productive collaboration, and is certainly not something I'd expect to see from someone putting themselves forward as a potential project maintainer.

jonatack commented:
My goodness, that's saddening to read. The only goal should be to improve Bitcoin. Let's please remember to assume good intentions, focus on the code and improving Bitcoin and leave the other stuff behind. Thank you!

vasild commented:
> > So, what happened? The relaytxes and minfeefilter fields were always present, and documented as such, in the output of the getpeerinfo RPC.
> 
> This isn't true. **None** of the fields in `getpeerinfo` were documented as always present until the 'optional' fields were explicitly documented in #22798. Prior to that PR, there was no explicit guarantee that any of the fields would be present.

Alright, by "documented" I meant that since it is not documented as optional, then it must be mandatory, but I was not aware that optional fields were introduced later (#22798). Thanks for bringing it up. This changes the perspective here.

> This was in response to a review comment about the GUI, not the RPC.

In the RPC case you referred to the GUI comment: [As above for the gui peer console](https://github.com/bitcoin/bitcoin/pull/21160#discussion_r598567367).

Anyway, I was just trying to clarify the situation here, not misinterpret, gotcha or shame. I am having coronavirus and should probably not be at the computer today at all, upsetting people. Will be more careful with my wording in the future.

Thank you!

1440000bytes commented:
Concept ACK

ajtowns commented on src/bitcoin-cli.cpp:
```diff
@@ -642,7 +642,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "  recv     Time since last message received from the peer, in seconds\n"
         "  txn      Time since last novel transaction received from the peer and accepted into our mempool, in minutes\n"
-        "           \"*\" - the peer requested we not relay transactions to it (relaytxes is false)\n"
+        "           \"*\" - the peer did not request that we relay transactions to it (relaytxes is false)\n"
```
I don't think this change is more accurate. `relaytxes` is false only when either a peer explicitly requests tx relay to be disabled (via setting the `fRelay` flag added in BIP 37 to false), or when we decide that an outgoing connection should be block-relay-only, in which case we ignore any request from it to relay transactions. Saying "whether we will relay txs with this peer" might be more accurate though.

jonatack replied:
Thanks, "whether we relay transactions with this peer" SGTM.

jonatack replied:
Updated both the getpeerinfo and the -netinfo help docs.

ajtowns commented on src/bitcoin-cli.cpp:
```diff
@@ -478,7 +478,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
             const int8_t network_id{NetworkStringToId(network)};
             if (network_id == UNKNOWN_NETWORK) continue;
             const bool is_outbound{!peer["inbound"].get_bool()};
-            const bool is_tx_relay{peer["relaytxes"].isNull() ? true : peer["relaytxes"].get_bool()};
```
`bitcoin-cli` could be being used with an older version of `bitcoind`, no? I think `get_bool` treats null as `false`, but I think `getInt` will throw. Probably not an issue if only unreleased versions of `bitcoind` treated these values as optional though.

MarcoFalke replied:
I don't think we support arbitrary version combinations of bitcoin-cli and the rpc server? The only official supported combination is the one where both are built from the exact same commit hash.

jonatack replied:
Right, for the moment the newly frequently missing fields are not in a release.

`get_bool` and `getInt` both throw on null with `error: JSON value of type null is not of expected type {bool|number}`, which can be seen with a snippet like this added to a CLI or RPC.

```cpp
const UniValue x = UniValue();
assert (x.isNull());
// auto y = x.get_bool();        // uncomment to throw
// auto z = x.getInt<int64_t>(); // uncomment to throw
```

jonatack replied:
> `bitcoin-cli` could be being used with an older version of `bitcoind`, no?

Circling back to this, as I have heard in the past from people using a recent -netinfo with nodes that have been up since an earlier release. Updated the minimum server check in -netinfo from v21 to v22, the last version that should be backward compatible (due to bitcoin core adding the addr_* fields in v22 used in -netinfo) and dropped only the relaytxes null check and not the addr_* ones to maintain that backward compatibilty.

ajtowns reviewed (commented):
I guess I'm not convinced either way about what the right behaviour is here -- if a connection is still being setup, it seems more useful to indicate that via `getpeerinfo` or `-netinfo` than just treating it as if it will be like most connections are. On the other hand, that's a bunch of effort for unusual special cases, so :shrug: ...

maflcko reviewed (commented):
While there are some fields that can obviously be made non-optional (such as the fee filter), because they can be absent on the wire even long after the connection has been spun up (thus require a default value anyway), there are other fields where I think it would be problematic to make them non-optional. For example, the service flags can not change after the handshake, so returning different service flags for the same peer id in different RPC invocations might cause issues in callers that assume they can't change. I see that you didn't change service flags in this pull, but the same should hold for txrelay, unless the user has bloom filters enabled.

So overall I don't think this can be "sold" as a bugfix of some kind, but instead should be done by evaluating each field on a case-by-case basis whether it makes more sense to be optional or not. The behaviour change should then be motivated based on that rationale and not to restore some underdocumented/misdocumented interface of a past release.

luke-jr commented on src/net_processing.cpp:
```diff
@@ -1459,50 +1459,52 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
-    stats.m_starting_height = peer->m_starting_height;
+    const PeerRef peer = GetPeerRef(nodeid);
+    if (peer == nullptr) {
+        return stats;
```
Concept ACK to returning partial data rather than none at all.

luke-jr commented on src/qt/rpcconsole.cpp:
```diff
@@ -1192,31 +1193,15 @@ void RPCConsole::updateDetailWidget()
-        ui->peerRelayTxes->setText(stats->nodeStateStats.m_relay_txs ? ts.yes : ts.no);
-    }
-
+    ui->peerServices->setText(GUIUtil::formatServicesStr(stats->nodeStateStats.m_services));
```
Concept NACK to displaying incorrect data when data is not available.

jonatack replied:
Done for the GUI
```diff
--- a/src/qt/guiutil.cpp
+++ b/src/qt/guiutil.cpp
@@ -748,7 +748,7 @@ QString formatServicesStr(quint64 mask)
     if (strList.size())
         return strList.join(", ");
     else
-        return QObject::tr("None");
+        return QObject::tr("Unknown");
 }
```
and left (for now) the RPC getpeerinfo behavior as it is currently, i.e. unchanged
```
    "services": "0000000000000000",
    "servicesnames": [
    ],
```

jonatack commented:
Rebased and updated to take review feedback by @ajtowns and @luke-jr and for the merge of #25717, maintaining the tristate `presynced_headers` behavior. The Clang-tidy CI errors are unrelated (https://github.com/bitcoin/bitcoin/pull/25963 should fix the Clang tidy one).

Thanks for the input, everyone. I'm agnostic to how we describe this change, as can be seen in the PR title and description history. I've re-updated both.  I agree with evaluating each field on a case-by-case basis and it's what I've tried to do.

History:
- the getpeerinfo rpc was added in 2012 in 1006f0707e34f8903f247195dabd86243ae61f05 by Jeff Garzik
- the `relaytxes` field was added to getpeerinfo in 2015 in 08843ed99843078 by Peter Todd
- the `minfeefilter` field was added to getpeerinfo in 2018 in 5778bf95d94f4d1 by @ajtowns  

Before adding the "optional" mentions to the RPC help docs, I believe we would often mention that fields were optional in the description, see 5ba829e12e99f11 for an example.

IIUC the only behavior change here where there may be varying opinions is to keep `relaytxes` always present.

My thinking is that this is the 8th year it has always been present, e.g. in a majority of our releases (~15) until now. That long-established behavior is a contract I would prefer to maintain, absent a compelling reason not to do so, in which case it might be good to warn our user space about the change in a release note.

It is true that both (a) the `getpeerinfo` fields order fixup and (b) keeping `relaytxes` and `minfeefilter` always present could be done separately for v24 with a small patch in `src/rpc/net.cpp` rather than in this larger change. 

Regarding these last two points, I'll leave it to reviewers to weigh in on which way to go.

jnewbery commented:
> Before adding the "optional" mentions to the RPC help docs, I believe we would often mention that fields were optional in the description, see https://github.com/bitcoin/bitcoin/commit/5ba829e12e99f119df56cab422f827b9be03fe57 for an example.

You're linking to documentation that *you* added to a single RPC. By comparison look at #22798, which adds over 120 optional type arguments to existing rpcs. A large majority of "optional" type arguments were not documented as such until September 2021/v0.23. The optional type itself was not added until #17809 in 2020, many years after these return fields were included in the RPC (and not documented as optional).

> That long-established behavior is a contract

No. If it's not documented, it's not a contract. Anyone writing clients using the RPC interface would have needed to make the clients accepting of fields being missing, since that's always been possible for a large number of the results fields in a large number of the RPCs.

I agree with Marco that it makes sense for some fields to be non-optional, so some of the changes here some beneficial, but please stick to the facts in describing this change.

jonatack commented:
Thanks @jnewbery. I'm agnostic to what word is used to describe the 8 years of `relaytxes` behavior. I think the question is whether reflecting that the peer's behavior is still being set up is a sufficient reason to change it for v24. When in doubt I would maintain the long-standing behavior but ok to update depending on consensus.

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -114,7 +114,7 @@ static RPCHelpMan getpeerinfo()
                         {RPCResult::Type::STR, "SERVICE_NAME", "the service name if it is recognised"}
                     }},
-                    {RPCResult::Type::BOOL, "relaytxes", /*optional=*/true, "Whether peer has asked us to relay transactions to it"},
+                    {RPCResult::Type::BOOL, "relaytxes", "Whether we relay transactions to this peer"},
```
in cdad951abf73650a9b49dd038ecbcc96ebf9bc0f: I think the doxygen for `m_relay_txs` should also be updated for the same reason. Otherwise it is not clear why you are changing this.

jonatack replied:
Good idea, done.

MarcoFalke replied:
> done

No?

jonatack replied:
Thanks, pushed

jonatack commented:
Rebased for the CI fixes and updated per @MarcoFalke and @ajtowns feedback (thanks).

> FWIW, I think you'd be better off justifying this on its own terms regardless of recent/ancient historical behaviour.

Updated the pull title/description and the commit messages. 

> I think `relaytxes: false` meaning "we currently won't relay txs to/from this peer" is clear, even if that might change after version negotiation. I could see `-1` or `CAmount::max()` being better defaults for `minfeefilter` if relaying isn't happening though, since that way you could interpret it correctly without needing to also look at `relaytxes`...

Makes sense, could see a -1 minfeefilter default for the reason you gave. Unsure with respect to API stability so deferring for a follow-up unless there is a preference that it be done here.

brunoerg commented on src/node/interfaces.cpp:
```diff
@@ -168,29 +168,25 @@ class NodeImpl : public Node
     bool getNodesStats(NodesStats& stats) override
```
Not specifically related to this PR (I guess): Wouldn't be better if we return `NodesStats` here rather than a bool?

brunoerg commented on src/node/interfaces.cpp:
```diff
@@ -168,29 +168,25 @@ class NodeImpl : public Node
     bool getNodesStats(NodesStats& stats) override
```
Couldn't we change this function to return `NodeStats` instead of `bool`. Seems odd the first thing we do here is `clear` the `stats`.

jonatack replied:
Good observation! Done.

jonatack committed:
interfaces, qt: always provide NodeStats

jonatack commented:
Rebased and ready for review.

brunoerg reviewed (approved):
crACK 4b10c63dc2f5c894c07adc12926862783cda99b1

Code seems good for me and noticed my suggestion has been addressed in one of the latest pushes.

naumenkogs commented:
I will halt a thorough code review for now while it seems like there is a non-trivial high-level ("concept") discussion of the subject.
This probably also means we should not rush it into the immediate release. There is certainly a risk of breaking some software, but the risks of merging something poor now could be worse (the amount of code here is non-trivial too).

As for my personal high-level opinion, I'm not yet convinced following the former behavior ("bugfix") is preferential (especially while not everyone argees it is a bugfix).

Please consider this my humble opinion as I haven't reviewed the history of these changes as close as other devs in the PR.

vasild commented on src/qt/rpcconsole.cpp:
```diff
@@ -1192,31 +1193,15 @@ void RPCConsole::updateDetailWidget()
-        ui->peerRelayTxes->setText(stats->nodeStateStats.m_relay_txs ? ts.yes : ts.no);
-    }
-
+    ui->peerServices->setText(GUIUtil::formatServicesStr(stats->nodeStateStats.m_services));
```
What about making `stats->nodeStateStats.m_services` `std::optional` and here do `x.has_value() ? ... : "Unknown")`? Otherwise we seem to be abusing a real value `NODE_NONE` (no services) as a dummy "we don't know/has no value".

jonatack replied:
Done

vasild reviewed (commented):
Posting review mid-way, I reviewed just the first commit 8c8773dc8744e58150fd09c145edaae80dd5ab4c.

I find it confusing to have `NodesStats`, `CNodeStats`, `CNodeStateStats` and `CNodeCombinedStats`.

Maybe changing

```cpp
using NodesStats = std::vector<std::pair<CNodeStats, CNodeStateStats>>;
```

to

```cpp
using CNodesCombinedStats = std::vector<CNodeCombinedStats>;
```

would alleviate it a bit?

jonatack commented on src/net_processing.cpp:
```diff
@@ -1546,56 +1549,57 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
-    stats.m_ping_wait = ping_wait;
     stats.m_addr_processed = peer->m_addr_processed.load();
     stats.m_addr_rate_limited = peer->m_addr_rate_limited.load();
     stats.m_addr_relay_enabled = peer->m_addr_relay_enabled.load();
     {
         LOCK(peer->m_headers_sync_mutex);
-        if (peer->m_headers_sync) {
-            stats.presync_height = peer->m_headers_sync->GetPresyncHeight();
-        }
+        stats.m_presync_height = peer->m_headers_sync ? peer->m_headers_sync->GetPresyncHeight() : -1;
```
I believe this branch maintains the current behavior for this field (correct me if I'm wrong), as it was member-initialized to -1 in the CNodeStateStats struct but only returned in the RPC if the peer was set up.

jonatack commented on src/net_processing.cpp:
```diff
@@ -1546,56 +1549,57 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
     // the caller can immediately detect that this is happening.
-    auto ping_wait{0us};
     if ((0 != peer->m_ping_nonce_sent) && (0 != peer->m_ping_start.load().count())) {
-        ping_wait = GetTime<std::chrono::microseconds>() - peer->m_ping_start.load();
+        stats.m_ping_wait = GetTime<std::chrono::microseconds>() - peer->m_ping_start.load();
     }
```
SGTM

jonatack replied:
Actually, we are only interested in non-zero values for pingwait, so it would be simpler not to do this. What do you think?

jonatack replied:
Reverted this change in the latest push at b5255ca75f.

vasild replied:
Since the trend in this PR is to use `std::optional` it is odd to use the magic value `0` for `m_ping_wait`. Consider this change, which I think should be possible to be amended in bbd680d23e `p2p, rpc, qt: always provide CNodeStateStats, update getpeerinfo and gui` (not tested):

```diff
diff --git i/src/net_processing.h w/src/net_processing.h
index 2d7068f4c8..50fbe679a2 100644
--- i/src/net_processing.h
+++ w/src/net_processing.h
@@ -26,13 +26,13 @@ static const bool DEFAULT_PEERBLOCKFILTERS = false;
 static const int DISCOURAGEMENT_THRESHOLD{100};

 struct CNodeStateStats {
     std::optional<int> m_sync_height;
     std::optional<int> m_common_height;
     std::optional<int> m_starting_height;
-    std::chrono::microseconds m_ping_wait{0us};
+    std::optional<std::chrono::microseconds> m_ping_wait;
     std::vector<int> m_height_in_flight;
     std::optional<bool> m_relay_txs{false};
     CAmount m_fee_filter_received{0};
     uint64_t m_addr_processed = 0;
     uint64_t m_addr_rate_limited = 0;
     bool m_addr_relay_enabled{false};
diff --git i/src/qt/guiutil.cpp w/src/qt/guiutil.cpp
index 10e1ea169a..6562308052 100644
--- i/src/qt/guiutil.cpp
+++ w/src/qt/guiutil.cpp
@@ -750,13 +750,13 @@ QString formatServicesStr(quint64 mask)
     else
         return QObject::tr("Unknown");
 }

 QString formatPingTime(std::chrono::microseconds ping_time)
 {
-    return (ping_time == std::chrono::microseconds::max() || ping_time == 0us) ?
+    return ping_time == std::chrono::microseconds::max() ?
         QObject::tr("N/A") :
         QObject::tr("%1 ms").arg(QString::number((int)(count_microseconds(ping_time) / 1000), 10));
 }

 QString formatTimeOffset(int64_t nTimeOffset)
 {
diff --git i/src/qt/rpcconsole.cpp w/src/qt/rpcconsole.cpp
index 49402b0264..088608397f 100644
--- i/src/qt/rpcconsole.cpp
+++ w/src/qt/rpcconsole.cpp
@@ -1194,13 +1194,13 @@ void RPCConsole::updateDetailWidget()
     }
     ui->peerMappedAS->setText(stats->nodeStats.m_mapped_as != 0 ? QString::number(stats->nodeStats.m_mapped_as) : ts.na);
     ui->peerServices->setText(stats->nodeStateStats.m_services.has_value() ? GUIUtil::formatServicesStr(stats->nodeStateStats.m_services.value()) : ts.unknown);
     ui->peerSyncHeight->setText(stats->nodeStateStats.m_sync_height.has_value() ? QString("%1").arg(stats->nodeStateStats.m_sync_height.value()) : ts.unknown);
     ui->peerCommonHeight->setText(stats->nodeStateStats.m_common_height.has_value() ? QString("%1").arg(stats->nodeStateStats.m_common_height.value()) : ts.unknown);
     ui->peerHeight->setText(stats->nodeStateStats.m_starting_height.has_value() ? QString::number(stats->nodeStateStats.m_starting_height.value()) : ts.unknown);
-    ui->peerPingWait->setText(GUIUtil::formatPingTime(stats->nodeStateStats.m_ping_wait));
+    ui->peerPingWait->setText(stats->nodeStateStats.m_ping_wait.has_value() ? GUIUtil::formatPingTime(stats->nodeStateStats.m_ping_wait.value()) : QObject::tr("N/A"));
     ui->peerAddrRelayEnabled->setText(stats->nodeStateStats.m_addr_relay_enabled ? ts.yes : ts.no);
     ui->peerAddrProcessed->setText(QString::number(stats->nodeStateStats.m_addr_processed));
     ui->peerAddrRateLimited->setText(QString::number(stats->nodeStateStats.m_addr_rate_limited));
     ui->peerRelayTxes->setText(stats->nodeStateStats.m_relay_txs ? ts.yes : ts.no);
     ui->peersTabRightPanel->show();
 }
diff --git i/src/rpc/net.cpp w/src/rpc/net.cpp
index c688bd658f..bbfd88980d 100644
--- i/src/rpc/net.cpp
+++ w/src/rpc/net.cpp
@@ -213,14 +213,14 @@ static RPCHelpMan getpeerinfo()
         if (stats.m_last_ping_time > 0us) {
             obj.pushKV("pingtime", Ticks<SecondsDouble>(stats.m_last_ping_time));
         }
         if (stats.m_min_ping_time < std::chrono::microseconds::max()) {
             obj.pushKV("minping", Ticks<SecondsDouble>(stats.m_min_ping_time));
         }
-        if (statestats.m_ping_wait > 0s) {
-            obj.pushKV("pingwait", Ticks<SecondsDouble>(statestats.m_ping_wait));
+        if (statestats.m_ping_wait.has_value()) {
+            obj.pushKV("pingwait", Ticks<SecondsDouble>(statestats.m_ping_wait.value()));
         }
         obj.pushKV("version", stats.nVersion);
         // Use the sanitized form of subver here, to avoid tricksy remote peers from
         // corrupting or modifying the JSON output by putting special characters in
         // their ver message.
         obj.pushKV("subver", stats.cleanSubVer);
```

jonatack commented on src/qt/guiutil.cpp:
```diff
@@ -748,7 +748,7 @@ QString formatServicesStr(quint64 mask)
         return strList.join(", ");
     else
-        return QObject::tr("None");
+        return QObject::tr("Unknown");
```
Done, back to "None" here and using "Unknown" in `src/qt/rpcconsole.cpp` per https://github.com/bitcoin/bitcoin/pull/25923#discussion_r963695601, and which continues to satisfy https://github.com/bitcoin/bitcoin/pull/25923#discussion_r956495436.

jonatack committed:
refactor: remove NodesStats type

as it can only be used in a couple of instances rather than in all
of them, thereby providing little benefit and increasing confusion

jonatack commented on src/net_processing.cpp:
```diff
@@ -1546,56 +1549,57 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
     {
         LOCK(cs_main);
         const CNodeState* state = State(nodeid);
-        if (state == nullptr)
-            return false;
-        stats.nSyncHeight = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
-        stats.nCommonHeight = state->pindexLastCommonBlock ? state->pindexLastCommonBlock->nHeight : -1;
+        if (state == nullptr) {
+            return stats;
```
I've added an explanatory comment. The idea is to return sensible defaults where it makes sense and simplifies the code (both ours and that of API clients). For instance, the "addr_processed" and "addr_rate_limited" fields are member-initialized to 0 and may as well be returned during set up, as the peer has processed and limited no addresses yet. Likewise, returning an "inflight" field with an empty array seems just as logical and makes life simpler for us and for clients, than sometimes not returning the field in the RPC response. Can update if reviewers disagree.

jonatack commented:
Thank you for the review @vasild! I've hopefully addressed your feedback.

> I find it confusing to have `NodesStats`, `CNodeStats`, `CNodeStateStats` and `CNodeCombinedStats` ... would alleviate it a bit?

Done.

fanquake commented:
p2p_ping is now failing:
```bash
222/251 - p2p_ping.py failed, Duration: 1 s
stdout:
2022-09-14T10:06:23.895000Z TestFramework (INFO): Initializing test directory /tmp/cirrus-ci-build/ci/scratch/test_runner/test_runner_₿_🏃_20220914_095106/p2p_ping_26
2022-09-14T10:06:24.156000Z TestFramework (INFO): Check that ping is sent after connection is established
2022-09-14T10:06:24.262000Z TestFramework (INFO): Reply without nonce cancels ping
2022-09-14T10:06:24.315000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 133, in main
    self.run_test()
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/p2p_ping.py", line 67, in run_test
    self.check_peer_info(pingtime=None, minping=None, pingwait=None)
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/p2p_ping.py", line 48, in check_peer_info
    assert_equal(stats.pop('pingwait', None), pingwait)
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 56, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(0 == None)
```

jonatack commented:
Thank you for reviewing @brunoerg and @vasild! Diff since your reviews: `git diff 4b10c63 c5a3587`

> p2p_ping is now failing

Yes, fixed.

jonatack commented:
> If this is stripped to the bare minimum needed, I think it will have higher chance of making it to 24.0

Done in a separate PR without the refactoring here; the latter can indeed be done later.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -1546,56 +1546,60 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
-            return false;
-        stats.nSyncHeight = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
-        stats.nCommonHeight = state->pindexLastCommonBlock ? state->pindexLastCommonBlock->nHeight : -1;
+        if (state == nullptr) {
```
Maybe I'm missing the context, but when would we expect to trigger this?
If this is called directly by user, why does it make sense to return defaults, rather than explicitly state the node is not found.

jonatack replied:
Thanks for having a look.  This can be seen e.g. in the GUI Peers tab details area when the next three blocks fields display "Unknown" during the peer setup phase.

The idea is to return sensible defaults where it makes sense and potentially simplifies the code (both ours and that of API clients). For instance, the "addr_processed" and "addr_rate_limited" fields are member-initialized to 0 and may as well be returned during set up, as the peer has processed and limited no addresses yet. Likewise, returning an "inflight" field with an empty array seems just as logical and makes life simpler for us and for clients, than sometimes not returning the field in the RPC response. For "pingwait" we are only interested in non-zero values so a default of zero makes sense. For "minfeefilter" we need a default, and so on. 

Will update the documentation here as follows:
```diff
         if (state == nullptr) {
-            // Return member-initialized sensible defaults for the fields where
-            // it makes sense and simplifies the code.
+            // Peer is still being set up; return member-initialized sensible defaults
+            // for the node state fields where it makes sense and simplifies the code.
             return stats;
```

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -1546,56 +1546,60 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
+    }
     stats.their_services = peer->m_their_services;
-    stats.m_starting_height = peer->m_starting_height;
+    if (peer->m_starting_height != -1) {
```
This is handling the two ways our code looks at `starting_height` unknown or something, which we create in b4bf7611d718e361dc7dea849550b3728da79136. Why not having one way? Switch to a new default if you wish, but having two different ones seems suboptimal.

jonatack replied:
I didn't do that yet as it would extend the refactoring from `CNodeStateStats` into the `Peer` class to change `m_starting_height` from `std::atomic` to `std::optional` with a mutex, but might still do it once this PR becomes refactoring-only (the other changes have been unbundled).

jonatack replied:
> I didn't do that yet as it would extend the refactoring from `CNodeStateStats` into the `Peer` class to change `m_starting_height` from `std::atomic` to `std::optional` with a mutex, but might still do it once this PR becomes refactoring-only (the other changes have been unbundled).

Started doing this in latest (WIP) push.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -1546,56 +1546,60 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
-        if (peer->m_headers_sync) {
-            stats.presync_height = peer->m_headers_sync->GetPresyncHeight();
-        }
+        stats.presync_height = peer->m_headers_sync ? peer->m_headers_sync->GetPresyncHeight() : -1;
```
in b4bf7611d718e361dc7dea849550b3728da79136

This looks like a legacy (w.r.t this PR) approach where we don't rely on defaults. Why not?

naumenkogs replied:
So the new behavior is the following:
- `nullptr` if peer==nullptr
- `-1` if peer->m_headers_sync == nullptr
- a real value otherwise

Is this supposed to be straightforward to the user?

jonatack replied:
It's not intended to be new behavior; will re-examine :)

jonatack committed:
p2p, refactor: make Peer#m_starting_height data member std::optional

naumenkogs commented on src/net_processing.h:
```diff
@@ -24,18 +26,18 @@ static const bool DEFAULT_PEERBLOCKFILTERS = false;
     std::vector<int> vHeightInFlight;
-    bool m_relay_txs;
-    CAmount m_fee_filter_received;
+    bool m_relay_txs{false};
```
I know you adjust the definition of this flag in later commits. I kinda still think `optional` before we know for sure would be better.

jonatack replied:
There seems to be agree on making it optional; will do.

jonatack replied:
Done in latest (WIP) push.

jonatack commented:
Rebased and WIP push (likely still need to add a mutex for `m_starting_height`), dropping the non-refactoring commits that have been merged in a separate pull.

vasild commented on src/net_processing.cpp:
```diff
@@ -2866,7 +2870,7 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
         // Headers message had its maximum size; the peer may have more headers.
         if (MaybeSendGetHeaders(pfrom, GetLocator(pindexLast), peer)) {
             LogPrint(BCLog::NET, "more getheaders (%d) to end to peer=%d (startheight:%d)\n",
-                    pindexLast->nHeight, pfrom.GetId(), peer.m_starting_height);
+                     pindexLast->nHeight, pfrom.GetId(), peer.m_starting_height.value());
```
If `m_starting_height` has no value the call to `value()` will throw an exception. I do not feel comfortable either assuming that it will always have value here (also considering future changes) or that throwing an exception is ok. Some of this code is surrounded by `try/catch` higher up, but we may move pieces of code around, possibly moving this printout elsewhere. Maybe do it more defensively:

```cpp
LogPrint(BCLog::NET, "more getheaders (%d) to end to peer=%d%s\n",
         pindexLast->nHeight,
         pfrom.GetId(),
         peer.m_starting_height.has_value() ? strprintf(" (startheight:%d)", peer.m_starting_height.value()) : "");
```

(here and in the other printouts)

vasild reviewed (commented):
Approach ACK 25fa0d5a6c11c79b3a2b60c4bba05a45f4c65d6a

Edit, just not to forget:

> likely still need to add a mutex for `m_starting_height`

or `std::atomic<std::optional<int>> m_starting_height`

Also consider: https://github.com/bitcoin/bitcoin/pull/25923#discussion_r963656857

jonatack commented on src/net_processing.h:
```diff
@@ -24,18 +26,18 @@ static const bool DEFAULT_PEERBLOCKFILTERS = false;
+    std::optional<int> m_starting_height;
+    std::chrono::microseconds m_ping_wait{0us};
+    std::vector<int> m_height_in_flight;
+    std::optional<bool> m_relay_txs{false};
```
Thanks @vasild! That was an oversight in my last WIP push, will update.

vasild commented on src/node/interfaces.cpp:
```diff
@@ -168,29 +168,25 @@ class NodeImpl : public Node
-                        std::get<1>(node_stats) =
-                            m_context->peerman->GetNodeStateStats(std::get<0>(node_stats).nodeid, std::get<2>(node_stats));
-                    }
+            TRY_LOCK(::cs_main, lockMain);
```
Is it not the same in `master`?

mzumsande replied:
I think in master, we return `fNodeStateStatsAvailable==false` and don't update the fields if we can't get the lock ([code](https://github.com/bitcoin/bitcoin/blob/6863ad79a65842504ab6f5584fac3d1de7ecf87e/src/qt/rpcconsole.cpp#L1200-L1201)) instead of updating them with default-values. `fNodeStateStatsAvailable` is being removed here.

mzumsande reviewed (commented):
Some observations re: the RPC changes (I think the reporting to the GUI should be discussed separately, see my comment above):

* As far as I can see, before this PR, the only situation in which `CNodeStateStats` was not available (`fStateStats==false`) is a race condition during peer disconnection, because `getpeerinfo` does not fetch `CNodeStats` and `CNodeStateStats` atomically - the node may have removed the peer in between the two calls. But this should not be possible during peer initialization because of the order in which the objects `CNode`, `Peer` and `CNodeState` are initialised.

* I think that this behavior during disconnection is unintended, a bug that should be fixed. One way to fix it would be to not report an incomplete peer info in this case (#26515), another possibility might be to prevent the node from disconnecting a peer in between these two calls, i.e. change the way `getpeerinfo` fetches data - which might also fix the related issue that data from `CNodeStateStats` might be always present but change after fetching `CNodeStats` data, leading to an inconsistent response.

* As a result of this, there currently seems to be no _legitimate_ reason why we should fail to fetch an entire `CNodeStateStats` object when we have a `CNodeStats` object, or why fields from `CNodeStateStats` need to be any more optional than field from `CNodeStats`. During the lifetime of a connection, all three internal objects `CNode`, `Peer` and `CNodeState` should always be available.

* It is an unrelated issue that some of the `getpeerinfo` fields (whichever internal object they may belong to) are not known during peer initialization because our peer hasn't given us enough info yet - the old strategy has been to return default-initialized values, such as `-1`, but include the field. This PR makes many of these fields optional, by leaving them out of the `getpeerinfo` response - so it changes `getpeerinfo` behavior during startup. This should become very clear when this PR gets rebased wrt master and the newly added test in #26519 needs to be adjusted.

* I don't have a strong opinion on whether this is desirable (are there any general recommendation for this, or examples from other RPCs?), however note that some fields (e.g. `presynced_headers=-1`) are still left default-initialized, while others aren't. I also think that if we make the fields truly optional, it might be good to mention in the RPC help in which situations they are not reported.

fanquake commented:
What is the status of this given recently merged changes, and the commentary above?

jonatack commented:
I'm evaluating the scope, starting from simply updating the statestats struct with std::optionals on up to the other changes.

fanquake commented:
> I'm evaluating the scope,

Ok. Moved to draft for now, given it's currently a work in progress. Feel free to undraft when you've decided on a scope, rebased, answered outstanding queries etc.
